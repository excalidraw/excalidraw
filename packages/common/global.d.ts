/// <reference types="vite/client" />
import "@excalidraw/excalidraw/global";
import "@excalidraw/excalidraw/css";
