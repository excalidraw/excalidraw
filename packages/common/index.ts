export * from "./src/binary-heap";
export * from "./src/colors";
export * from "./src/constants";
export * from "./src/keys";
export * from "./src/points";
export * from "./src/promise-pool";
export * from "./src/random";
export * from "./src/url";
export * from "./src/utils";
