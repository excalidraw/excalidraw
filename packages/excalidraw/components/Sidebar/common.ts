import React from "react";
import { AppState, SidebarName, SidebarTabName } from "../../types";

export type SidebarTriggerProps = {
  name: SidebarName;
  tab?: SidebarTabName;
  icon?: JSX.Element;
  children?: React.ReactNode;
  title?: string;
  className?: string;
  onToggle?: (open: boolean) => void;
  style?: React.CSSProperties;
};

export type SidebarProps<P = {}> = {
  name: SidebarName;
  children: React.ReactNode;
  /**
   * Called on sidebar open/close or tab change.
   */
  onStateChange?: (state: AppState["openSidebar"]) => void;
  /**
   * supply alongside `docked` prop in order to make the Sidebar user-dockable
   */
  onDock?: (docked: boolean) => void;
  docked?: boolean;
  className?: string;
  // NOTE sidebars we use internally inside the editor must have this flag set.
  // It indicates that this sidebar should have lower precedence over host
  // sidebars, if both are open.
  /** @private internal */
  __fallback?: boolean;
} & P;

export type SidebarPropsContextValue = Pick<
  SidebarProps,
  "onDock" | "docked"
> & { onCloseRequest: () => void; shouldRenderDockButton: boolean };

export const SidebarPropsContext =
  React.createContext<SidebarPropsContextValue>({} as SidebarPropsContextValue);
