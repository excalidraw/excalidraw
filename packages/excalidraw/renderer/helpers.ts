import type { StaticCanvasAppState, AppState } from "../types";

import type { StaticCanvasRenderConfig } from "../scene/types";

import { THEME, THEME_FILTER } from "../constants";

export const fillCircle = (
  context: CanvasRenderingContext2D,
  cx: number,
  cy: number,
  radius: number,
  stroke = true,
) => {
  context.beginPath();
  context.arc(cx, cy, radius, 0, Math.PI * 2);
  context.fill();
  if (stroke) {
    context.stroke();
  }
};

export const getNormalizedCanvasDimensions = (
  canvas: HTMLCanvasElement,
  scale: number,
): [number, number] => {
  // When doing calculations based on canvas width we should used normalized one
  return [canvas.width / scale, canvas.height / scale];
};

export const bootstrapCanvas = ({
  canvas,
  scale,
  normalizedWidth,
  normalizedHeight,
  theme,
  isExporting,
  viewBackgroundColor,
}: {
  canvas: HTMLCanvasElement;
  scale: number;
  normalizedWidth: number;
  normalizedHeight: number;
  theme?: AppState["theme"];
  isExporting?: StaticCanvasRenderConfig["isExporting"];
  viewBackgroundColor?: StaticCanvasAppState["viewBackgroundColor"];
}): CanvasRenderingContext2D => {
  const context = canvas.getContext("2d")!;

  context.setTransform(1, 0, 0, 1, 0, 0);
  context.scale(scale, scale);

  if (isExporting && theme === THEME.DARK) {
    context.filter = THEME_FILTER;
  }

  // Paint background
  if (typeof viewBackgroundColor === "string") {
    const hasTransparence =
      viewBackgroundColor === "transparent" ||
      viewBackgroundColor.length === 5 || // #RGBA
      viewBackgroundColor.length === 9 || // #RRGGBBA
      /(hsla|rgba)\(/.test(viewBackgroundColor);
    if (hasTransparence) {
      context.clearRect(0, 0, normalizedWidth, normalizedHeight);
    }
    context.save();
    context.fillStyle = viewBackgroundColor;
    context.fillRect(0, 0, normalizedWidth, normalizedHeight);
    context.restore();
  } else {
    context.clearRect(0, 0, normalizedWidth, normalizedHeight);
  }

  return context;
};
