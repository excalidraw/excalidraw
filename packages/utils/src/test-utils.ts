import { diffStringsUnified } from "jest-diff";

expect.extend({
  toCloselyEqualPoints(received, expected, precision) {
    if (!Array.isArray(received) || !Array.isArray(expected)) {
      throw new Error("expected and received are not point arrays");
    }

    const COMPARE = 1 / Math.pow(10, precision || 2);
    const pass = expected.every(
      (point, idx) =>
        Math.abs(received[idx]?.[0] - point[0]) < COMPARE &&
        Math.abs(received[idx]?.[1] - point[1]) < COMPARE,
    );

    if (!pass) {
      return {
        message: () => ` The provided array of points are not close enough.

${diffStringsUnified(
  JSON.stringify(expected, undefined, 2),
  JSON.stringify(received, undefined, 2),
)}`,
        pass: false,
      };
    }

    return {
      message: () => `expected ${received} to not be close to ${expected}`,
      pass: true,
    };
  },
});
