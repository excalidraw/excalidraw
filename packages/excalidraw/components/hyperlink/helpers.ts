import { MIME_TYPES } from "../../constants";
import type { Bounds } from "../../element/bounds";
import { getElementAbsoluteCoords } from "../../element/bounds";
import { hitElementBoundingBox } from "../../element/collision";
import type {
  ElementsMap,
  NonDeletedExcalidrawElement,
} from "../../element/types";
import { rotate } from "../../math";
import { DEFAULT_LINK_SIZE } from "../../renderer/renderElement";
import type { AppState, Point, UIAppState } from "../../types";

export const EXTERNAL_LINK_IMG = document.createElement("img");
EXTERNAL_LINK_IMG.src = `data:${MIME_TYPES.svg}, ${encodeURIComponent(
  `<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="#1971c2" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-external-link"><path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"></path><polyline points="15 3 21 3 21 9"></polyline><line x1="10" y1="14" x2="21" y2="3"></line></svg>`,
)}`;

export const getLinkHandleFromCoords = (
  [x1, y1, x2, y2]: Bounds,
  angle: number,
  appState: Pick<UIAppState, "zoom">,
): Bounds => {
  const size = DEFAULT_LINK_SIZE;
  const linkWidth = size / appState.zoom.value;
  const linkHeight = size / appState.zoom.value;
  const linkMarginY = size / appState.zoom.value;
  const centerX = (x1 + x2) / 2;
  const centerY = (y1 + y2) / 2;
  const centeringOffset = (size - 8) / (2 * appState.zoom.value);
  const dashedLineMargin = 4 / appState.zoom.value;

  // Same as `ne` resize handle
  const x = x2 + dashedLineMargin - centeringOffset;
  const y = y1 - dashedLineMargin - linkMarginY + centeringOffset;

  const [rotatedX, rotatedY] = rotate(
    x + linkWidth / 2,
    y + linkHeight / 2,
    centerX,
    centerY,
    angle,
  );
  return [
    rotatedX - linkWidth / 2,
    rotatedY - linkHeight / 2,
    linkWidth,
    linkHeight,
  ];
};

export const isPointHittingLinkIcon = (
  element: NonDeletedExcalidrawElement,
  elementsMap: ElementsMap,
  appState: AppState,
  [x, y]: Point,
) => {
  const threshold = 4 / appState.zoom.value;
  const [x1, y1, x2, y2] = getElementAbsoluteCoords(element, elementsMap);
  const [linkX, linkY, linkWidth, linkHeight] = getLinkHandleFromCoords(
    [x1, y1, x2, y2],
    element.angle,
    appState,
  );
  const hitLink =
    x > linkX - threshold &&
    x < linkX + threshold + linkWidth &&
    y > linkY - threshold &&
    y < linkY + linkHeight + threshold;
  return hitLink;
};

export const isPointHittingLink = (
  element: NonDeletedExcalidrawElement,
  elementsMap: ElementsMap,
  appState: AppState,
  [x, y]: Point,
  isMobile: boolean,
) => {
  if (!element.link || appState.selectedElementIds[element.id]) {
    return false;
  }
  if (
    !isMobile &&
    appState.viewModeEnabled &&
    hitElementBoundingBox(x, y, element, elementsMap)
  ) {
    return true;
  }
  return isPointHittingLinkIcon(element, elementsMap, appState, [x, y]);
};
