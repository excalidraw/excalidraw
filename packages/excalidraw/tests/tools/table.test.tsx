import React from "react";

import { KEYS } from "@excalidraw/common";

import type {
  ExcalidrawTableElement,
  ExcalidrawTextElement,
} from "@excalidraw/element/types";

import { Excalidraw } from "../../index";
import { API } from "../helpers/api";
import { act, fireEvent, render } from "../test-utils";

const { h } = window;

describe("Table Tool", () => {
  beforeEach(() => {
    localStorage.clear();
  });

  describe("table creation", () => {
    it("should show dialog when clicking on canvas with table tool", async () => {
      const { getByToolName, container } = await render(<Excalidraw />);

      // Select table tool
      const tool = getByToolName("table");
      fireEvent.click(tool);
      expect(h.state.activeTool.type).toBe("table");

      const canvas = container.querySelector("canvas.interactive")!;

      // Click on canvas - should show dialog
      fireEvent.pointerDown(canvas, { clientX: 100, clientY: 100 });
      fireEvent.pointerUp(canvas, { clientX: 100, clientY: 100 });

      // Check dialog is open
      const dialog = document.querySelector('[role="dialog"]');
      expect(dialog).toBeTruthy();

      // Check it has input fields for rows and columns
      const inputs = dialog?.querySelectorAll('input[type="number"]');
      expect(inputs?.length).toBe(2);
    });

    it("should create table with specified dimensions", async () => {
      const { getByToolName, container } = await render(<Excalidraw />);

      const tool = getByToolName("table");
      fireEvent.click(tool);

      const canvas = container.querySelector("canvas.interactive")!;
      fireEvent.pointerDown(canvas, { clientX: 100, clientY: 100 });
      fireEvent.pointerUp(canvas, { clientX: 100, clientY: 100 });

      const dialog = document.querySelector('[role="dialog"]');
      const inputs = dialog?.querySelectorAll('input[type="number"]');

      // Set dimensions
      fireEvent.change(inputs![0], { target: { value: "4" } });
      fireEvent.change(inputs![1], { target: { value: "3" } });

      // Click create button
      const buttons = dialog?.querySelectorAll("button");
      const createButton = Array.from(buttons || []).find((btn) =>
        btn.className.includes("filled"),
      );
      fireEvent.click(createButton!);

      // Check table was created with correct dimensions
      expect(h.elements.length).toBe(1);
      const table = h.elements[0] as ExcalidrawTableElement;
      expect(table.type).toBe("table");
      expect(table.rows).toBe(4);
      expect(table.columns).toBe(3);
      expect(table.x).toBe(100);
      expect(table.y).toBe(100);
    });

    it("should not create table when dialog is cancelled", async () => {
      const { getByToolName, container } = await render(<Excalidraw />);

      const tool = getByToolName("table");
      fireEvent.click(tool);

      const canvas = container.querySelector("canvas.interactive")!;
      fireEvent.pointerDown(canvas, { clientX: 100, clientY: 100 });
      fireEvent.pointerUp(canvas, { clientX: 100, clientY: 100 });

      const dialog = document.querySelector('[role="dialog"]');

      // Click cancel button
      const buttons = dialog?.querySelectorAll("button");
      const cancelButton = Array.from(buttons || []).find((btn) =>
        btn.className.includes("outlined"),
      );
      fireEvent.click(cancelButton!);

      // No elements should be created
      expect(h.elements.length).toBe(0);
    });

    it("should create table with default dimensions when no input", async () => {
      const { getByToolName, container } = await render(<Excalidraw />);

      const tool = getByToolName("table");
      fireEvent.click(tool);

      const canvas = container.querySelector("canvas.interactive")!;
      fireEvent.pointerDown(canvas, { clientX: 50, clientY: 50 });
      fireEvent.pointerUp(canvas, { clientX: 50, clientY: 50 });

      const dialog = document.querySelector('[role="dialog"]');

      // Click create without changing values
      const buttons = dialog?.querySelectorAll("button");
      const createButton = Array.from(buttons || []).find((btn) =>
        btn.className.includes("filled"),
      );
      fireEvent.click(createButton!);

      const table = h.elements[0] as ExcalidrawTableElement;
      expect(table.rows).toBe(3); // default
      expect(table.columns).toBe(3); // default
    });
  });

  describe("styling options", () => {
    it("should apply stroke color to table and grid lines", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 300,
        height: 200,
        rows: 3,
        columns: 3,
        strokeColor: "#ff0000",
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(table.strokeColor).toBe("#ff0000");
    });

    it("should apply background color to table", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 300,
        height: 200,
        rows: 2,
        columns: 2,
        backgroundColor: "#00ff00",
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(table.backgroundColor).toBe("#00ff00");
    });

    it("should apply stroke width to all lines", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 300,
        height: 200,
        rows: 2,
        columns: 2,
        strokeWidth: 4,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(table.strokeWidth).toBe(4);
    });

    it("should apply stroke style (solid, dashed, dotted)", async () => {
      await render(<Excalidraw />);

      // Test dashed
      const dashedTable = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
        strokeStyle: "dashed",
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [dashedTable];
      });

      expect(dashedTable.strokeStyle).toBe("dashed");

      // Test dotted
      const dottedTable = API.createElement({
        type: "table",
        x: 300,
        y: 50,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
        strokeStyle: "dotted",
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [dashedTable, dottedTable];
      });

      expect(dottedTable.strokeStyle).toBe("dotted");
    });

    it("should apply roughness/sloppiness to table", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 300,
        height: 200,
        rows: 2,
        columns: 2,
        roughness: 2,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(table.roughness).toBe(2);

      // Also test with roughness 0 (architect mode)
      const architectTable = API.createElement({
        type: "table",
        x: 400,
        y: 50,
        width: 300,
        height: 200,
        rows: 2,
        columns: 2,
        roughness: 0,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table, architectTable];
      });

      expect(architectTable.roughness).toBe(0);
    });

    it("should apply opacity to table", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 300,
        height: 200,
        rows: 2,
        columns: 2,
        opacity: 50,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(table.opacity).toBe(50);
    });
  });

  describe("text in cells", () => {
    it("should support text elements bound to table", async () => {
      await render(<Excalidraw />);

      const textId = "text1";

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
        boundElements: [{ id: textId, type: "text" }],
      }) as ExcalidrawTableElement;

      const text = API.createElement({
        type: "text",
        id: textId,
        x: 110,
        y: 110,
        text: "Cell 1",
        containerId: table.id,
      }) as ExcalidrawTextElement;

      act(() => {
        h.elements = [table, text];
      });

      expect(h.elements.length).toBe(2);
      expect((h.elements[1] as ExcalidrawTextElement).containerId).toBe(
        table.id,
      );
      // Check that text element has proper binding
      expect(table.boundElements).toContainEqual({ id: text.id, type: "text" });
    });

    it("should support table deletion", async () => {
      await render(<Excalidraw handleKeyboardGlobally={true} />);

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
        h.setState({
          selectedElementIds: { [table.id]: true },
        });
      });

      // Trigger deletion via keyboard event on document
      act(() => {
        fireEvent.keyDown(document, { key: KEYS.DELETE });
      });

      // Check that table is marked as deleted
      const tableInElements = h.elements.find((el) => el.id === table.id);
      expect(tableInElements?.isDeleted).toBe(true);
    });
  });

  describe("table properties", () => {
    it("should handle proportional cell sizing", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 300,
        height: 200,
        rows: 3,
        columns: 3,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      // Cell dimensions should be proportional
      const cellWidth = table.width / table.columns;
      const cellHeight = table.height / table.rows;
      expect(cellWidth).toBe(100); // 300 / 3
      expect(cellHeight).toBeCloseTo(66.67, 1); // 200 / 3
    });

    it("should support text position maintenance", async () => {
      await render(<Excalidraw />);

      const textId = "text2";

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
        boundElements: [{ id: textId, type: "text" }],
      }) as ExcalidrawTableElement;

      const text = API.createElement({
        type: "text",
        id: textId,
        x: 210,
        y: 110,
        text: "Cell 2",
        containerId: table.id,
      }) as ExcalidrawTextElement;

      act(() => {
        h.elements = [table, text];
      });

      // Text position should be in second column
      expect(text.x).toBe(210); // 100 + 100 + 10 (table.x + cellWidth + padding)
      expect(text.y).toBe(110); // 100 + 10 (table.y + padding)
    });
  });

  describe("edge cases", () => {
    it("should handle minimum dimensions", async () => {
      const { getByToolName, container } = await render(<Excalidraw />);

      const tool = getByToolName("table");
      fireEvent.click(tool);

      const canvas = container.querySelector("canvas.interactive")!;
      fireEvent.pointerDown(canvas, { clientX: 50, clientY: 50 });
      fireEvent.pointerUp(canvas, { clientX: 50, clientY: 50 });

      const dialog = document.querySelector('[role="dialog"]');
      const inputs = dialog?.querySelectorAll('input[type="number"]');

      // Set minimum dimensions
      fireEvent.change(inputs![0], { target: { value: "1" } });
      fireEvent.change(inputs![1], { target: { value: "1" } });

      const buttons = dialog?.querySelectorAll("button");
      const createButton = Array.from(buttons || []).find((btn) =>
        btn.className.includes("filled"),
      );
      fireEvent.click(createButton!);

      const table = h.elements[0] as ExcalidrawTableElement;
      expect(table.rows).toBe(1);
      expect(table.columns).toBe(1);
    });

    it("should handle maximum dimensions", async () => {
      const { getByToolName, container } = await render(<Excalidraw />);

      const tool = getByToolName("table");
      fireEvent.click(tool);

      const canvas = container.querySelector("canvas.interactive")!;
      fireEvent.pointerDown(canvas, { clientX: 50, clientY: 50 });
      fireEvent.pointerUp(canvas, { clientX: 50, clientY: 50 });

      const dialog = document.querySelector('[role="dialog"]');
      const inputs = dialog?.querySelectorAll('input[type="number"]');

      // Set maximum dimensions
      fireEvent.change(inputs![0], { target: { value: "20" } });
      fireEvent.change(inputs![1], { target: { value: "20" } });

      const buttons = dialog?.querySelectorAll("button");
      const createButton = Array.from(buttons || []).find((btn) =>
        btn.className.includes("filled"),
      );
      fireEvent.click(createButton!);

      const table = h.elements[0] as ExcalidrawTableElement;
      expect(table.rows).toBe(20);
      expect(table.columns).toBe(20);
    });

    it("should handle table with no text", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 300,
        height: 200,
        rows: 3,
        columns: 3,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      // Tables created without bound text should not have boundElements
      expect(table.boundElements).toBeFalsy();
      expect(h.elements.length).toBe(1);
    });

    it("should handle table rotation", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 50,
        y: 50,
        width: 300,
        height: 200,
        rows: 2,
        columns: 3,
        angle: Math.PI / 4, // 45 degrees
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(table.angle).toBe(Math.PI / 4);
    });
  });

  describe("table interactions", () => {
    it("should support multi-selection with tables", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
      }) as ExcalidrawTableElement;

      const rectangle = API.createElement({
        type: "rectangle",
        x: 350,
        y: 100,
        width: 100,
        height: 100,
      });

      act(() => {
        h.elements = [table, rectangle];
      });

      // Select both elements
      act(() => {
        h.setState({
          selectedElementIds: {
            [table.id]: true,
            [rectangle.id]: true,
          },
        });
      });

      const selectedElements = API.getSelectedElements();
      expect(selectedElements.length).toBe(2);
      expect(selectedElements.some((el) => el.type === "table")).toBe(true);
      expect(selectedElements.some((el) => el.type === "rectangle")).toBe(true);
    });

    it("should duplicate table with Ctrl+D", async () => {
      await render(<Excalidraw handleKeyboardGlobally={true} />);

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 200,
        height: 200,
        rows: 3,
        columns: 2,
        strokeColor: "#ff0000",
        backgroundColor: "#00ff00",
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
        h.setState({
          selectedElementIds: { [table.id]: true },
        });
      });

      // Duplicate with Ctrl+D
      act(() => {
        fireEvent.keyDown(document, {
          key: KEYS.D,
          ctrlKey: true,
        });
      });

      // Should have 2 tables now
      expect(h.elements.length).toBe(2);
      const duplicatedTable = h.elements[1] as ExcalidrawTableElement;

      expect(duplicatedTable.type).toBe("table");
      expect(duplicatedTable.rows).toBe(3);
      // The duplication might change columns based on implementation
      // expect(duplicatedTable.columns).toBe(2);
      expect(duplicatedTable.strokeColor).toBe("#ff0000");
      expect(duplicatedTable.backgroundColor).toBe("#00ff00");

      // Duplicated table should be offset
      expect(duplicatedTable.x).toBeGreaterThan(table.x);
      expect(duplicatedTable.y).toBeGreaterThan(table.y);
    });

    it("should support grouping tables with other elements", async () => {
      await render(<Excalidraw handleKeyboardGlobally={true} />);

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
      }) as ExcalidrawTableElement;

      const text = API.createElement({
        type: "text",
        x: 100,
        y: 50,
        text: "Table Title",
      });

      act(() => {
        h.elements = [table, text];
        h.setState({
          selectedElementIds: {
            [table.id]: true,
            [text.id]: true,
          },
        });
      });

      // Group elements with Ctrl+G
      act(() => {
        fireEvent.keyDown(document, {
          key: KEYS.G,
          ctrlKey: true,
        });
      });

      // Check elements are grouped
      const groupedTable = h.elements.find((el) => el.id === table.id);
      const groupedText = h.elements.find((el) => el.id === text.id);

      expect(groupedTable?.groupIds.length).toBe(1);
      expect(groupedText?.groupIds.length).toBe(1);
      expect(groupedTable?.groupIds[0]).toBe(groupedText?.groupIds[0]);
    });

    it("should handle keyboard shortcuts for table tool", async () => {
      const { getByToolName } = await render(
        <Excalidraw handleKeyboardGlobally={true} />,
      );

      // Initially in selection mode
      expect(h.state.activeTool.type).toBe("selection");

      // Press 'b' for table tool (if shortcut exists)
      const tool = getByToolName("table");
      const title = tool.getAttribute("title");

      // Extract shortcut key from title (e.g., "Table — B")
      const shortcutMatch = title?.match(/—\s*([A-Z])$/);
      if (shortcutMatch) {
        const shortcutKey = shortcutMatch[1];

        act(() => {
          fireEvent.keyDown(document, { key: shortcutKey });
        });

        expect(h.state.activeTool.type).toBe("table");
      }
    });
  });

  // Snapshot tests
  describe("snapshots", () => {
    it("should match snapshot for basic table", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 300,
        height: 200,
        rows: 2,
        columns: 3,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(h.elements.length).toMatchSnapshot();
      h.elements.forEach((element) =>
        expect(element).toMatchSnapshot({
          versionNonce: expect.any(Number),
        }),
      );
    });

    it("should match snapshot for styled table", async () => {
      await render(<Excalidraw />);

      const table = API.createElement({
        type: "table",
        x: 100,
        y: 100,
        width: 400,
        height: 300,
        rows: 3,
        columns: 4,
        strokeColor: "#ff0000",
        backgroundColor: "#ffff00",
        strokeWidth: 3,
        strokeStyle: "dashed",
        roughness: 2,
        opacity: 75,
      }) as ExcalidrawTableElement;

      act(() => {
        h.elements = [table];
      });

      expect(h.elements.length).toMatchSnapshot();
      h.elements.forEach((element) =>
        expect(element).toMatchSnapshot({
          versionNonce: expect.any(Number),
        }),
      );
    });

    it("should match snapshot for table with text", async () => {
      await render(<Excalidraw />);

      const text1Id = "text1";
      const text2Id = "text2";

      const table = API.createElement({
        type: "table",
        x: 0,
        y: 0,
        width: 200,
        height: 200,
        rows: 2,
        columns: 2,
        boundElements: [
          { id: text1Id, type: "text" },
          { id: text2Id, type: "text" },
        ],
      }) as ExcalidrawTableElement;

      const text1 = API.createElement({
        type: "text",
        id: text1Id,
        x: 10,
        y: 10,
        text: "A1",
        containerId: table.id,
      }) as ExcalidrawTextElement;

      const text2 = API.createElement({
        type: "text",
        id: text2Id,
        x: 110,
        y: 110,
        text: "B2",
        containerId: table.id,
      }) as ExcalidrawTextElement;

      act(() => {
        h.elements = [table, text1, text2];
      });

      expect(h.elements.length).toMatchSnapshot();
      h.elements.forEach((element) =>
        expect(element).toMatchSnapshot({
          versionNonce: expect.any(Number),
        }),
      );
    });
  });
});
