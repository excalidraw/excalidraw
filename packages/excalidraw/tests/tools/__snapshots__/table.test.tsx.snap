// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Tool > snapshots > should match snapshot for basic table 1`] = `1`;

exports[`Table Tool > snapshots > should match snapshot for basic table 2`] = `
{
  "angle": 0,
  "backgroundColor": "transparent",
  "boundElements": null,
  "columns": 3,
  "customData": undefined,
  "fillStyle": "solid",
  "frameId": null,
  "groupIds": [],
  "height": 200,
  "id": "id33",
  "index": "a0",
  "isDeleted": false,
  "link": null,
  "locked": false,
  "opacity": 100,
  "roughness": 1,
  "roundness": {
    "type": 3,
  },
  "rows": 2,
  "seed": 1,
  "strokeColor": "#1e1e1e",
  "strokeStyle": "solid",
  "strokeWidth": 2,
  "type": "table",
  "updated": 1,
  "version": 2,
  "versionNonce": Any<Number>,
  "width": 300,
  "x": 100,
  "y": 100,
}
`;

exports[`Table Tool > snapshots > should match snapshot for styled table 1`] = `1`;

exports[`Table Tool > snapshots > should match snapshot for styled table 2`] = `
{
  "angle": 0,
  "backgroundColor": "#ffff00",
  "boundElements": null,
  "columns": 4,
  "customData": undefined,
  "fillStyle": "solid",
  "frameId": null,
  "groupIds": [],
  "height": 300,
  "id": "id34",
  "index": "a0",
  "isDeleted": false,
  "link": null,
  "locked": false,
  "opacity": 75,
  "roughness": 2,
  "roundness": {
    "type": 3,
  },
  "rows": 3,
  "seed": 1,
  "strokeColor": "#ff0000",
  "strokeStyle": "dashed",
  "strokeWidth": 3,
  "type": "table",
  "updated": 1,
  "version": 2,
  "versionNonce": Any<Number>,
  "width": 400,
  "x": 100,
  "y": 100,
}
`;

exports[`Table Tool > snapshots > should match snapshot for table with text 1`] = `3`;

exports[`Table Tool > snapshots > should match snapshot for table with text 2`] = `
{
  "angle": 0,
  "backgroundColor": "transparent",
  "boundElements": [
    {
      "id": "text1",
      "type": "text",
    },
    {
      "id": "text2",
      "type": "text",
    },
  ],
  "columns": 2,
  "customData": undefined,
  "fillStyle": "solid",
  "frameId": null,
  "groupIds": [],
  "height": 200,
  "id": "id35",
  "index": "a0",
  "isDeleted": false,
  "link": null,
  "locked": false,
  "opacity": 100,
  "roughness": 1,
  "roundness": {
    "type": 3,
  },
  "rows": 2,
  "seed": 1,
  "strokeColor": "#1e1e1e",
  "strokeStyle": "solid",
  "strokeWidth": 2,
  "type": "table",
  "updated": 1,
  "version": 2,
  "versionNonce": Any<Number>,
  "width": 200,
  "x": 0,
  "y": 0,
}
`;

exports[`Table Tool > snapshots > should match snapshot for table with text 3`] = `
{
  "angle": 0,
  "autoResize": true,
  "backgroundColor": "transparent",
  "boundElements": null,
  "containerId": "id35",
  "customData": undefined,
  "fillStyle": "solid",
  "fontFamily": 5,
  "fontSize": 20,
  "frameId": null,
  "groupIds": [],
  "height": 100,
  "id": "text1",
  "index": "a1",
  "isDeleted": false,
  "lineHeight": "1.25000",
  "link": null,
  "locked": false,
  "opacity": 100,
  "originalText": "A1",
  "roughness": 1,
  "roundness": {
    "type": 3,
  },
  "seed": 1,
  "strokeColor": "#1e1e1e",
  "strokeStyle": "solid",
  "strokeWidth": 2,
  "text": "A1",
  "textAlign": "left",
  "type": "text",
  "updated": 1,
  "version": 2,
  "versionNonce": Any<Number>,
  "verticalAlign": "top",
  "width": 100,
  "x": 10,
  "y": 10,
}
`;

exports[`Table Tool > snapshots > should match snapshot for table with text 4`] = `
{
  "angle": 0,
  "autoResize": true,
  "backgroundColor": "transparent",
  "boundElements": null,
  "containerId": "id35",
  "customData": undefined,
  "fillStyle": "solid",
  "fontFamily": 5,
  "fontSize": 20,
  "frameId": null,
  "groupIds": [],
  "height": 100,
  "id": "text2",
  "index": "a2",
  "isDeleted": false,
  "lineHeight": "1.25000",
  "link": null,
  "locked": false,
  "opacity": 100,
  "originalText": "B2",
  "roughness": 1,
  "roundness": {
    "type": 3,
  },
  "seed": 1,
  "strokeColor": "#1e1e1e",
  "strokeStyle": "solid",
  "strokeWidth": 2,
  "text": "B2",
  "textAlign": "left",
  "type": "text",
  "updated": 1,
  "version": 2,
  "versionNonce": Any<Number>,
  "verticalAlign": "top",
  "width": 100,
  "x": 110,
  "y": 110,
}
`;
